pyro/distributions/sylvester.py

```diff
@@ -0,0 +1,136 @@
+    orthogonality of :math:`Q` is enforced by representing it as the product of Householder transformations
+
+    Together with `TransformedDistribution` it provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> hsf = SylvesterFlow(10, count_transforms=4)
+    >>> pyro.module("my_hsf", hsf)  # doctest: +SKIP
+    >>> hsf_dist = dist.TransformedDistribution(base_dist, [hsf])
+    >>> hsf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse of this transform does not possess an analytical solution and is left unimplemented. However,
+    the inverse is cached when the forward operation is called during sampling, and so samples drawn using
+    Sylvester flow can be scored.
+
+    References:
+
+    Rianne van den Berg, Leonard Hasenclever, Jakub M. Tomczak, Max Welling. Sylvester Normalizing Flows for
+    Variational Inference. In proceedings of The 34th Conference on Uncertainty in Artificial Intelligence
+    (UAI 2018).
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+
+    def __init__(self, input_dim, count_transforms=1):
+        super(SylvesterFlow, self).__init__(input_dim, count_transforms)
+
+        # Create parameters for Sylvester transform
+        self.R_dense = nn.Parameter(torch.Tensor(input_dim, input_dim))
+        self.S_dense = nn.Parameter(torch.Tensor(input_dim, input_dim))
+        self.R_diag = nn.Parameter(torch.Tensor(input_dim))
+        self.S_diag = nn.Parameter(torch.Tensor(input_dim))
+        self.b = nn.Parameter(torch.Tensor(input_dim))
+
+        # Register masks and indices
+        triangular_mask = torch.triu(torch.ones(input_dim, input_dim), diagonal=1)
+        self.register_buffer('triangular_mask', triangular_mask)
+
+        self._cached_logDetJ = None
+        self.tanh = nn.Tanh()
+        self.reset_parameters2()
+
+    # Derivative of hyperbolic tan
+    def dtanh_dx(self, x):
+        return 1. - self.tanh(x).pow(2)
+
+    # Construct upper diagonal R matrix
+    def R(self):
+        return self.R_dense * self.triangular_mask + torch.diag(self.tanh(self.R_diag))
+
+    # Construct upper diagonal S matrix
+    def S(self):
+        return self.S_dense * self.triangular_mask + torch.diag(self.tanh(self.S_diag))
+
+    # Construct orthonomal matrix using Householder flow
+    def Q(self):
+        u = self.u()
+        partial_Q = torch.eye(self.input_dim) - 2. * torch.ger(u[0], u[0])
+
+        for idx in range(1, self.count_transforms):
+            partial_Q = torch.matmul(partial_Q, torch.eye(self.input_dim) - 2. * torch.ger(u[idx], u[idx]))
+
+        return partial_Q
+
+    # NOTE: self.u_unnormed is initialized in parent class
+    def reset_parameters2(self):
```

Comment:
is there a way to do this by overriding`reset_parameters` so you don't have 2 of them floating around?

Reply:
Overriding `reset_parameters` would make things difficult. If I overrode it, I would need to change the position of `super(SylvesterFlow, self).__init__(input_dim, count_transforms)`  so that it didn't try to initialize `self.R_dense` etc. before they were created. But creating `self.R_dense` before the initializer is called would introduce a bug because the initializer for `torch.nn.Module` won't have been called yet, and those parameters won't be added to `self._modules`. Any solution to this would be much more complicated than just calling the method `reset_parameters2`